apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/S3DirectoryDao.java

```diff
@@ -0,0 +1,458 @@
+   *
+   * @param s3Key the S3 key
+   * @return equivalent file name
+   */
+  private String convertS3KeyToFileName(String s3Key) {
+    return s3Key.substring(s3DirectoryPath.length());
+  }
+
+  /**
+   * Reads a list of all files in the S3 bucket/directory.
+   *
+   * @return the object listing
+   */
+  private ObjectListing readS3ObjectListing() {
+    if (Strings.isNullOrEmpty(s3DirectoryPath)) {
+      return s3Client.listObjects(s3BucketName);
+    } else {
+      return s3Client.listObjects(s3BucketName, s3DirectoryPath);
+    }
+  }
+
+  /**
+   * Ensure that the directory path is either empty or ends with a slash character.
+   *
+   * @param directoryName s3 directory name
+   * @return normalized directory name
+   */
+  private static String normalizeDirectoryPath(String directoryName) {
+    if (directoryName == null) {
+      return "";
+    } else if (directoryName.isEmpty() || directoryName.endsWith("/")) {
+      return directoryName;
+    } else {
+      return directoryName + "/";
+    }
+  }
+
+  /**
+   * Convert a simple file name plus the file's eTag value into a {@link Path} referencing a file in
+   * our cache directory.
+   *
+   * @param fileName simple file name as exposed to clients
+   * @param eTag eTag value from S3 for the file
+   * @return file handle
+   */
+  @VisibleForTesting
+  Path cacheFilePath(String fileName, String eTag) {
+    String cacheFileName = DataFilePrefix + fileName + EtagSeparator + eTag + DataFileSuffix;
+    return Path.of(cacheDirectory.toString(), cacheFileName);
+  }
+
+  /**
+   * Convert a {@link S3ObjectSummary} into a {@link Path} referencing a file in our cache
+   * directory.
+   *
+   * @param summary {@link S3ObjectSummary} for object to store in cache
+   * @return file handle
+   */
+  private Path cacheFilePath(S3ObjectSummary summary) {
+    final var fileName = convertS3KeyToFileName(summary.getKey());
+    return cacheFilePath(fileName, summary.getETag());
+  }
+
+  /**
+   * Verifies that the given {@link Path} is one created by us to store an S3 object in our cache
+   * directory.
+   *
+   * @param path file to check
+   * @return true if file has a valid data file name
+   */
+  private boolean isCacheFile(Path path) {
+    final var fileName = path.getFileName().toString();
+    return Files.isRegularFile(path)
+        && fileName.startsWith(DataFilePrefix)
+        && fileName.endsWith(DataFileSuffix);
+  }
+
+  /**
+   * Creates a {@link ByteSource} for reading a cached file. If the file is a gzip file (name ends
+   * with .gz) the file will be automatically decompressed when a stream is opened. Otherwise the
+   * bytes will be returned as they appear in the file.
+   *
+   * @param fileName file name from the S3 object key
+   * @param path location of the cached file
+   * @return the byte source
+   */
+  private ByteSource createByteSourceForCachedFile(String fileName, Path path) {
+    var byteSource = MoreFiles.asByteSource(path);
```

Comment:
Is it an issue that this API is not considered stable?

Reply:
Yeah that just means that it might change in a future version of guava.  I don't think it's concerning.